File: customer-api-spa/types/bookings/ExcursionCardUI.ts
import { Uuid } from '../basic/Uuid'
import { LangItem } from '../LangItem'
import { ImageAPI } from './ImageAPI'
import { ExcursionTagItem } from './ExcursionTagItem'
import { PointUI } from '../geo/PointUI'
import { WeekDay } from '../WeekDay'
import { IntRange } from '../IntRange'
import { Amount } from './Amount'
import { PlannerExcursionDate } from '../planner/PlannerExcursionDates'
import { parseLocalDate, parseLocalTime } from '../DateRange'

export interface ExcursionCardUI {
  id: Uuid;
  excursionId: Uuid;
  lang: LangItem;
  name: string;
  description: string;
  clientTerms: string;
  paymentTerms: string;
  cancellationTerms: string;
  taxTerms: string;
  images: ImageAPI[];
  tags: ExcursionTagItem[];
  pickupPoint: PointUI;
  dates: string[];
  accommodationPax: number;
  startTime: string;
  duration: number;
  days: WeekDay[];
  age: IntRange;
  price: Amount;
}


export const orderByDate = (eds: PlannerExcursionDate[]) => (l: ExcursionCardUI, r: ExcursionCardUI) => {
  const leftDateString = eds.find(x => x.excursionOfferId === l.id)
  const rightDateString = eds.find(x => x.excursionOfferId === r.id)
  if (!leftDateString || !rightDateString){
    return 0
  }
  const leftDate = parseLocalDate(leftDateString.selected).valueOf()
  const rightDate = parseLocalDate(rightDateString.selected).valueOf()
  const compareDates = leftDate - rightDate
  if (compareDates < 0){
    return -1
  } else if (compareDates > 0) {
    return 1
  } else {
    const leftTime = parseLocalTime(l.startTime).valueOf()
    const rightTime = parseLocalTime(r.startTime).valueOf()
    return leftTime - rightTime
  }
}

File: excursions-protocol/src/main/scala/bookingtour/protocols/excursions/api/ExcursionCardUI.scala
package bookingtour.protocols.excursions.api

import java.time.{LocalDate, LocalTime}

import bookingtour.protocols.core._
import bookingtour.protocols.core.newtypes.quantities.{Hour, Pax, WeekDay}
import bookingtour.protocols.core.values.api.ImageAPI
import bookingtour.protocols.core.values.enumeration.{ExcursionTagItem, LangItem}
import bookingtour.protocols.core.values.{Amount, Ranges}
import bookingtour.protocols.excursions.newTypes.{ExcursionId, ExcursionOfferId}
import bookingtour.protocols.parties.api.PointUI
import cats.instances.all._
import derevo.cats.order
import derevo.circe.{decoder, encoder}
import derevo.derive

/**
  * © Alexey Toroshchin 2020.
  */
@derive(encoder, decoder, order)
final case class ExcursionCardUI(
    id: ExcursionOfferId,
    excursionId: ExcursionId,
    lang: LangItem,
    name: String,
    description: String,
    clientTerms: String,
    paymentTerms: String,
    cancellationTerms: String,
    taxTerms: String,
    images: List[ImageAPI],
    tags: List[ExcursionTagItem],
    pickupPoint: PointUI,
    dates: List[LocalDate],
    accommodationPax: Pax,
    startTime: LocalTime,
    duration: Hour,
    days: List[WeekDay],
    age: Ranges.Ints,
    price: Amount
)

object ExcursionCardUI {
  final type Id = ExcursionOfferId

  implicit final val itemR: ExcursionCardUI => Id = _.id

  implicit final val itemP: ExcursionCardUI => Int = _ => 0
}
